crypto/bn/bn_s390x.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2023-2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/bn.h"
+#include "crypto/s390x_arch.h"
+
+#ifdef S390X_MOD_EXP
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <asm/zcrypt.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <errno.h>
+
+static int s390x_mod_exp_hw(BIGNUM *r, const BIGNUM *a, const BIGNUM *p,
+                            const BIGNUM *m)
+{
+    struct ica_rsa_modexpo me;
+    unsigned char *buffer;
+    size_t size;
+    int res = 0;
+
+    if (OPENSSL_s390xcex == -1)
+        return 0;
+    size = BN_num_bytes(m);
+    buffer = OPENSSL_zalloc(4 * size);
+    if (buffer == NULL)
+        return 0;
+    me.inputdata = buffer;
+    me.inputdatalength = size;
+    me.outputdata = buffer + size;
+    me.outputdatalength = size;
+    me.b_key = buffer + 2 * size;
+    me.n_modulus = buffer + 3 * size;
+    if (BN_bn2binpad(a, me.inputdata, size) == -1 ||
+        BN_bn2binpad(p, me.b_key, size) == -1 ||
+        BN_bn2binpad(m, me.n_modulus, size) == -1)
+        goto dealloc;
+    if (ioctl(OPENSSL_s390xcex, ICARSAMODEXPO, &me) != -1) {
+        if (BN_bin2bn(me.outputdata, size, r) != NULL)
+            res = 1;
+    } else if (errno == EBADF) {
+        /*-
+         * In this cases, someone (e.g. a sandbox) closed the fd.
+         * Make sure to not further use this hardware acceleration.
+         */
+        (void)close(OPENSSL_s390xcex);
+        OPENSSL_s390xcex = -1;
+    }
```

Comment:
I am concerned about this being a race condition. Would it be OK to drop this case? Or at least drop the close() call assuming that the assignment is atomic.

Reply:
Actually, the `close` is not needed at all.  The file descriptor is already closed.  So this was just intended to optimize future calls into this function that directly fail and do not perform any computations that are anyway doomed to fail.  Removed the close.

Checked in the disassembly:  Storing `-1` to a global variable is as atomic as can be expected:  Load address of variable in register, then write `-1` to indirect-0 of register.  And since I ever only store `-1` to that variable (besides in the constructor), there is no store-store race.  As for load-store races, they do not cause problems in this scenario since the file descriptor is anyway closed leading to `errno == EBADF` which is the same result as issuing an `ioctl` on file descriptor `-1`.